cpp/src/groupby/new_groupby.cu

```diff
@@ -190,3 +192,230 @@ gdf_error gdf_group_by(gdf_column* in_key_columns[],
 
   return gdf_error_code;
 }
+
+
+
+// thrust::device_vector set to use rmmAlloc and rmmFree.
+template <typename T>
+using Vector = thrust::device_vector<T, rmm_allocator<T>>;
+
+
+namespace {
+
+
+/* --------------------------------------------------------------------------*/
+  /**
+   * @brief Given a set of columns, of which a subset of these are defined to be the key columns,
+   * all input data is assumed to already be sorted by these group by columns. 
+   * This function calculates a list of the start indices of each group
+   *
+   * @Param[in] num_data_cols The number of columns in the dataset (assumed to already be sorted)
+   * @Param[in] data_cols_in The input columns in the dataset
+   * @Param[in] num_groupby_cols  The number of key columns by
+   * @Param[in] The column indices of the input dataset that will be grouped by
+   * @Param[out] group_start_indices A column containing the starting indices of each group. Indices based off of new sort order. (needs to be pre-allocated)
+   * @Param[in] ctxt The context used to control how nulls are treated in a sort 
+   *
+   * @Returns gdf_error with error code on failure, otherwise GDF_SUCESS
+   */
+  /* ----------------------------------------------------------------------------*/
+gdf_error gdf_group_start_indices(gdf_size_type num_data_cols,
+                     gdf_column** data_cols_in,
+									   gdf_size_type num_groupby_cols,
+									   gdf_index_type * groupby_col_indices,
+									   gdf_column* group_start_indices,
+									   gdf_context* ctxt)
+{
+
+  int32_t nrows = data_cols_in[0]->size;
+  // setup for reduce by key  
+  bool have_nulls = false;
+  for (int i = 0; i < num_groupby_cols; i++) {
+    if (data_cols_in[i]->null_count > 0) {
+      have_nulls = true;
+      break;
+    }
+  }
+
+  Vector<void*> d_cols(num_groupby_cols); 
+  Vector<int> d_types(num_groupby_cols, 0);
+  void** d_col_data = d_cols.data().get();
+  int* d_col_types = d_types.data().get();
+
+  bool nulls_are_smallest = ctxt->flag_nulls_sort_behavior == 1;
+
+  int32_t* result_end;
```

Comment:
Don't hardcode types used for sizes or indices. Always use `gdf_size_type` or `gdf_index_type` instead to insulate your code from if those types ever change.
```suggestion
  gdf_index_type* result_end;
```